Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/triangulate_faces.h

```diff
@@ -93,126 +92,169 @@ class Triangulate_modifier
     return fh->info().is_external;
   }
 
-  void triangulate_face(face_descriptor f, PM& pmesh)
+  bool triangulate_face(face_descriptor f, PM& pmesh)
   {
-    typename Traits::Vector_3 normal =
-      Polygon_mesh_processing::compute_face_normal(f, pmesh);
 
-    P_traits cdt_traits(normal);
-    CDT cdt(cdt_traits);
 
-    // Halfedge_around_facet_circulator
-    typedef typename CDT::Vertex_handle Tr_Vertex_handle;
-    halfedge_descriptor start = halfedge(f, pmesh);
-    halfedge_descriptor h = start;
-    Tr_Vertex_handle previous, first;
-    do
+    typename Traits::Vector_3 normal =
+      Polygon_mesh_processing::compute_face_normal(f, pmesh);
+    if(normal == typename Traits::Vector_3(0,0,0))
+      return false;
+    std::size_t original_size = CGAL::halfedges_around_face(halfedge(f, pmesh), pmesh).size();
+    if(original_size == 4)
     {
-      Tr_Vertex_handle vh = cdt.insert(_vpmap[target(h, pmesh)]);
-      if (first == Tr_Vertex_handle()) {
-        first = vh;
+      halfedge_descriptor v0, v1, v2, v3;
+      v0 = halfedge(f, pmesh);
+      Point_ref p0 = _vpmap[target(v0, pmesh)];
+      v1 = next(v0, pmesh);
+      Point_ref p1 = _vpmap[target(v1, pmesh)];
+      v2 = next(v1, pmesh);
+      Point_ref p2 = _vpmap[target(v2, pmesh)];
+      v3 = next(v2, pmesh);
+      Point_ref p3 = _vpmap[target(v3, pmesh)];
+
+      //Chooses the diagonal that will maximize the cross product. The dot product allows to
+      // choose the right diagonal if the quad is not convex.
+      double p1p3= CGAL::cross_product(p2-p1,p3-p2) * CGAL::cross_product(p0-p3,p1-p0);
+      double p0p2= CGAL::cross_product(p1-p0,p1-p2) * CGAL::cross_product(p3-p2,p3-p0);
+
```

Comment:
The comment is wrong. The expressions `p1p3` and `p0p2` are not cross products, but scalar products of cross products.

I suggest something like the following lines:
> Choose the diagonal that will split the quad in two triangles that maximize the scalar product of of the un-normalized normals of the two triangles. The lengths of the un-normalized normals (computed using cross-products of two vectors) are proportional to the area of the triangles.
> Maximize the scalar product of the two normals will avoid skinny triangles, and will also taken into account the cosine of the angle between the two normals. In particular, if the two triangles are oriented in different directions, the scalar product will be negative.